DOCSP-49279: $convert stage LINQ method (#604)
mongodb/docs-csharp: source/fundamentals/linq.txt

   { "_id" : ObjectId(...), "name" : "Abaleh", "restaurant_id" : "50009096", "cuisine" : "Mediterranean", "address" : { ... }, "borough" : "Manhattan", "grades" : [...] }
   ...

.. _csharp-linq-convert:

$convert
~~~~~~~~

> You can use the ``Mql.Convert()`` method in a ``Select()`` projection to

Reviewer: The use of `Mql.Convert` is not limited only to `Select`, even though it could be the most common.

> You can use the ``Mql.Convert()`` method in a ``Select()`` projection to

Reviewer: I can see from the other operators in the page that usually we first introduce briefly the operator, and then show how to use it in LINQ. Maybe we should follow that pattern?

Author: I moved some sections around to distinguish between stages/operators

> convert a value from one type to a different specified type. You can
> use To learn more about conversion behavior and permitted conversions,

Reviewer 2: ```suggestion
convert a value from one type to a different specified type.
To learn more about conversion behavior and permitted conversions,
```

:manual:`$convert reference </reference/operator/aggregation/convert/>`
in the {+mdb-server+} manual.

The ``Convert()`` method takes the following parameters:

- Value to convert.

> - ``ConvertOptions<Tto>`` instance that specifies the type to convert to

Reviewer: Here we can omit `TTo`, as we did not introduce it before.

  and options. Some conversions require you to specify certain options, but
  you can also set options to handle errors or null values.

The following code converts the ``RestaurantId`` string values to
``int`` values in a projection by using LINQ:

.. code-block:: csharp
   
   var query = queryableCollection

>        .Select(r => Mql.Convert(r.RestaurantId, new ConvertOptions<int>()));

Author: Hi @papafe would you prefer if we showed a more complex example that uses the options?

Reviewer: I think the example is fine, it's difficult to make a more complex example using the dataset we have in this page. But at least you can show `OnError` or `OnNull`. So it could be:

```
   var query = queryableCollection
       .Select(r => Mql.Convert(r.RestaurantId, new ConvertOptions<int>( OnError = -1, OnNull = 0 )));
```

> The converted values in the output documents are stored under the original
> field name.

Reviewer 2: ```suggestion
The driver stores the converted values in the output documents under the original
field name.
```

   0
   1

.. _csharp-linq-convert:

$convert
~~~~~~~~

The ``$convert`` operator converts a value to a specified type. You can
use this operator to perform type conversions in stages such as
``$project``, ``$addFields``, and ``$set``.

In the driver, you can use the ``Mql.Convert()`` method to
convert a value from one type to a different specified type. To learn
more about conversion behavior and permitted conversions, see the
:manual:`$convert reference </reference/operator/aggregation/convert/>`
in the {+mdb-server+} manual.

The ``Convert()`` method takes the following parameters:

- Value to convert.
- ``ConvertOptions`` instance that specifies the type to convert to
  and options. Some conversions require you to specify certain options, but
  you can also set options to handle errors or null values.

The following code performs the following actions by using LINQ methods:

- Converts the ``RestaurantId`` string values to ``int`` values in a
  ``Select()`` projection
- Sets the returned value to ``-1`` if an error occurs during conversion
- Sets the returned value to ``0`` if the input value is ``null`` or missing

.. code-block:: csharp
   
   var query = queryableCollection

>        .Select(r => Mql.Convert(r.RestaurantId, new ConvertOptions<int>( OnError = -1, OnNull = 0 )));

Reviewer: ```suggestion
       .Select(r => Mql.Convert(r.RestaurantId, new ConvertOptions<int> { OnError = -1, OnNull = 0  }));
```
My bad, I used the wrong parenthesis in the example 🤦
